Add first solution hw2-2

Nigma-Ks opened the pull request.

Nigma-Ks committed:
Add first solution hw2-2

Nigma-Ks committed:
interface renamed

Nigma-Ks committed:
+null, hw2-1->hw2-2

Lozov-Petr commented on hw2_2/hw2-2/Program.cs:
```diff
@@ -0,0 +1,298 @@
+using System;
+using System.Collections.Generic;
+
+namespace hw2_2
```
В C# принято разбивать проект на несколько файлов: в каждом файле один класс.
Из этого бывают исключения (если имеем совсем маленькие однотипные классы), но это не наш случай.

Lozov-Petr replied:
В C# принято разбивать проект на несколько файлов: в каждом файле один класс.
Из этого бывают исключения (если имеем совсем маленькие однотипные классы), но это не наш случай.

Lozov-Petr commented on hw2_2/hw2-2/Program.cs:
```diff
@@ -0,0 +1,298 @@
+        static void Main(string[] args)
+        {
+            StackCalculatorOfReversePolishNotation calculator = new();
+            bool correctNotation;
```
Очень заранее объявлена переменная. Перенеси ближе к месту использования.

Lozov-Petr commented on hw2_2/hw2-2/Program.cs:
```diff
@@ -0,0 +1,298 @@
+                "1. Stack on array base\n" +
+                "2. Stack on list base\n");
+            Console.Write("write number 1 or 2: ");
+            string choice = Console.ReadLine();
```
```suggestion
            string choice = Console.ReadLine().Trim();
```
Что пробелы игнорировать.
Только вызывать надо чуть иначе, учитывая, что применяем не к строке, а к `string?`.

Lozov-Petr replied:
Тип результата `ReadLine()` - это `string?` и работать с ним надо соответственно.

Lozov-Petr commented on hw2_2/hw2-2/Program.cs:
```diff
@@ -0,0 +1,298 @@
+            }
+
+            Console.Write("Enter expression in reverse polish notation: ");
+            string expression = Console.ReadLine();
```
Снова не `string`, а `string?`.

Lozov-Petr commented on hw2_2/hw2-2/Program.cs:
```diff
@@ -0,0 +1,298 @@
+
+        private float[] ExpandArrayOnDefaultSize(float[] arrayToExpand)
+        {
+            float[] newExpandedArray = new float[arrayToExpand.Length + defaultArraySize];
```
Метод `Array.Resize` упростит изменение размера массива.

Lozov-Petr commented on hw2_2/hw2-2/Program.cs:
```diff
@@ -0,0 +1,298 @@
+    {
+        void Add(float value);
+
+        float Remove(out bool isEmpty);
```
```suggestion
        (float value, bool isEmpty) Remove();
```
Мы на паре договорились вот так несколько значений возвращать.
Если неясно, почему так лучше, спроси на паре.

Lozov-Petr commented on hw2_2/hw2-2/Program.cs:
```diff
@@ -0,0 +1,298 @@
+
+    public class StackCalculatorOfReversePolishNotation
+    {
+        public float delta = 0.0001f; //null comparison precision
```
```suggestion
        const float delta = 0.0001f; //null comparison precision
```
Если что-то является константой, нужно делать это константой.
Так и код понятнее, и защита от искушения таки изменить это значение.

Lozov-Petr commented on hw2_2/hw2-2/Program.cs:
```diff
@@ -0,0 +1,298 @@
+                StringSplitOptions.RemoveEmptyEntries); //if there more than one spaces
+        }
+
+        public float ToCalculate(string expessionInReversePolishNotation, IStack stack,
```
Не нужно возвращать через параметр, просто верни пару. А еще, можно вернуть `float?`, чтобы в случае отсутствия ответа не возвращать никакого ответа. Тогда и дополнительный результат типа `bool` возвращать не надо.

Lozov-Petr commented on hw2_2/hw2-2/Program.cs:
```diff
@@ -0,0 +1,298 @@
+                }
+                else
+                {
+                    if (parsedExpression[i] == "+")
```
Копипаста. Нужно объединить эти случаи в один.

Lozov-Petr commented on hw2_2/hw2-2/Program.cs:
Нет тестов

Lozov-Petr reviewed (changes requested):
2

Nigma-Ks committed:
1. Классы в разных файлах
2. string -> string?
3. возвращение не через параметр, сразу парой
4. убрала копипасту в калькуляторе
